### DISCUSSION TITLE
Re-run dependency output read per group in run-all

### DISCUSSION THREAD

**Discussion by kholisrag** [gruntwork-io/knowledge-base#549] [Help]
Hi,
is there any options to re-reun dependency output reading per group when we use run-all?

imagine that, we have multiple group when running run-all,
some dependency have mock_outputs,
I want to use real value in Group 2 rather than the mock_outputs after Group 1 truly applied,

is there possible?

---

<ins datetime="2022-08-31T10:57:51Z">
  <p><a href="https://support.gruntwork.io/hc/requests/109191">Tracked in ticket #109191</a></p>
</ins>

**Comment by zackproser** [comment]
Hi @kholisrag could you please provide a simple code configuration demonstrating the scenario you're describing?

**Reply by kholisrag** [reply]
will try that, seem still find a way to demonstrate that.

what I do is like : 

```
root
    shared
        unique-id
           terragrunt.hcl
    services
       serviceA
          terragrunt.hcl    
    unique-id.hcl
```

`uniqeu-id.hcl`
```
dependency "unique_id" {
  config_path = format("%s/%s/shared/unique-id", get_path_to_repo_root(), get_path_from_repo_root())

  mock_outputs = {
    id = "000000"
  }
}
```

`serviceA` `terragrunt.hcl`
```
locals {
  unique_id_vars    = read_terragrunt_config(find_in_parent_folders("unique-id.hcl"))
  unique_id             = local.unique_id_vars.dependency.unique_id.outputs.id
}

inputs {
  unique_id = local.unique_id
}

```

but based on 
https://terragrunt.gruntwork.io/docs/getting-started/configuration/#configuration-parsing-order

seem in `locals` parsed the outputs from the `read_terragrunt_config(find_in_parent_folders("unique-id.hcl"))`, from the mock output.
rather than waiting it to truly running the dependency first. cmiiw

**Reply by kholisrag** [reply]
basically what I want is to centralized the `dependency` groups in parent directory. then child `terragrunt.hcl` read that. rather than one by one adding the `dependency` in each child `terragrunt.hcl`

**Reply by kholisrag** [reply]
iirc I already try to use include, but because of `Limitations on accessing exposed config` in https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/#include
it failed.

**Reply by kholisrag** [reply]
imagine if have multiple group in the run-all apply, 
if previous group done, its better to get the output again, rather than using first output,

my problem will be solved, if I explicitly specify the `dependency` of all dependencies to the current `terragrunt.hcl` file, but it will be pain  in the ass, if we do that, so my take is to centralized the `dependency` specification, but then above concern happen.

**Reply by kholisrag** [reply]
@zackproser any difficulty when you reproduce?

**Comment by denis256** [comment]
hi,
I was thinking about a setup where `skip_outputs` option can be configured in each module.
Simplified example:
```
.
├── services
│   ├── serviceA-mock
│   │   ├── config.hcl      # configuration loaded in unique-id.hcl
│   │   ├── main.tf
│   │   └── terragrunt.hcl
│   └── serviceA-real-value
│       ├── config.hcl     # configuration loaded in unique-id.hcl
│       ├── main.tf
│       └── terragrunt.hcl
├── shared
│   └── unique-id
│       ├── main.tf
│       └── terragrunt.hcl
└── unique-id.hcl
```

```
# unique-id.hcl

locals {
  # read config file from each module to know in which one output is disabled
  config    = read_terragrunt_config("config.hcl")
}

dependency "unique_id" {
  config_path = format("%s/group-dependency-output-include/shared/unique-id", get_path_to_repo_root())
  # configure outputs skipping
  skip_outputs = "${local.config.locals.disable_outputs}"
  mock_outputs = {
    id = "000000"
  }
}

# services/serviceA-mock/terragrunt.hcl
include "unique_id" {
  path   = find_in_parent_folders("unique-id.hcl")
  merge_strategy = "deep"
}

inputs = {
  unique_id = dependency.unique_id.outputs.id
}

# services/serviceA-mock/config.hcl
locals {
  disable_outputs = true
}

# services/serviceA-real-value/config.hcl
locals {
  disable_outputs = false
}
```

On the submitted example I did a [similar setup](https://github.com/denis256/terragrunt-tests/tree/master/group-dependency-output) and indeed it uses mock outputs each time.


References:

https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/#dependency

https://github.com/denis256/terragrunt-tests/tree/master/group-dependency-output-include
